nixos/systemd-boot: Add `extraEntries` and `extraFiles` options

Enzime opened the pull request:
###### Motivation for this change

These options allow people to add their own EFI files and corresponding entries to `systemd-boot`.

I have also added `boot.loader.systemd-boot.netbootxyz` options to allow people to easily add `netboot.xyz` to their boot menu.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Enzime committed:
netbootxyz-efi: init at 2.0.53

Enzime committed:
nixos/systemd-boot: Support extra EFI entries

Enzime committed:
nixos/systemd-boot: Add option to add netboot.xyz

pennae commented on pkgs/tools/misc/netbootxyz-efi/default.nix:
```diff
@@ -0,0 +1,28 @@
+  pname = "netboot.xyz-efi";
+  version = "2.0.53";
+
+  src = fetchurl {
```
this could probably be the entire derivation, doesn't look like there's a need for a mkDerivation wrapper around the fetch

Enzime replied:
Fixed :+1:

pennae commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix:
```diff
@@ -125,6 +141,74 @@ in {
+    };
+
+    extraFiles = mkOption {
+      type = types.attrsOf types.path;
```
might be a good idea to add a safety net against path traversals in the keys of this set (eg disallow `/` prefix and `..` infix)

Enzime replied:
Fixed :+1:

pennae reviewed (commented):
is it feasible to add a nixos test to ensure the generated boot menu is functional and extras are cleaned up properly? looks pretty neat otherwise apart from a few nitpicks.

pennae commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix:
```diff
@@ -29,6 +29,22 @@ let
+      empty_file=$(mktemp)
+
+      ${concatStrings (mapAttrsToList (n: v: ''
+        ${pkgs.coreutils}/bin/install -Dp "${v}" "${efi.efiSysMountPoint}/${n}"
```
unlikely to cause an error in practice, but one more nit:
```suggestion
        ${pkgs.coreutils}/bin/install -Dp "${v}" "${efi.efiSysMountPoint}"/${escapeShellArg n}
```
also for the others.

Enzime replied:
Fixed :+1:

Enzime committed:
tests/systemd-boot: Add tests for `extraFiles`

pennae commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix:
```diff
@@ -148,15 +232,61 @@ in {
         message = "This kernel does not support the EFI boot stub";
       }
+      {
+        assertion = builtins.all (filename: !(hasInfix "/" filename)) (builtins.attrNames cfg.extraEntries);
+        message = "systemd-boot does not support entries within folders";
```
if possible assertions should give an indication where the error occured to make it easier to find. all of these would have to be split to generate one assertion per condition and thing checked to allow for that, eg with error messages like
```
boot.loader.systemd-boot.extraEntries.${lib.strings.escapeNixIdentifier attr} is invalid: entries within folders are not supported
```

Enzime replied:
Fixed :+1:

pennae commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix:
```diff
@@ -148,15 +232,61 @@ in {
     boot.loader.supportsInitrdSecrets = true;
 
+    boot.loader.systemd-boot.extraFiles = mkMerge [
+      # TODO: This is hard-coded to use the 64-bit EFI app, but it could probably
+      # be updated to use the 32-bit EFI app on 32-bit systems.  The 32-bit EFI
+      # app filename is BOOTIA32.efi.
```
you could check `stdenv.targetPlatform.is32bit` for this. maybe it's also worth adding an assert that `stdenv.targetPlatform.isx86` when x86-only items are enabled?

Enzime replied:
This was taken straight from the old code, I didn't make any changes to the `memtest86` stuff as I don't use it

pennae replied:
yup, did see it. was merely a suggestion for if you felt like addressing that one too while you're at it. :)

pennae reviewed (commented):
looks good now, just one thing we missed on the first go and one suggestion :)

pennae reviewed (approved):
tested locally, looks good!

Enzime committed:
nixos/nixos-enter: Don't passthru `TMPDIR`

`mktemp` tries to use the `TMPDIR` from `nixos-install` outside of the
`chroot` instead of `/tmp` inside the `chroot` and fails. For some
reason the `TMPDIR` is being passed through the `chroot` call.

I haven't tested if other environment variables are being passed through
that shouldn't be.

Enzime commented:
@pennae you might be interested in taking a look at the newest commit I just made. Found this bug when doing a fresh install of NixOS using this branch :smile:

Mic92 commented:
ping cc @pennae

pennae commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix:
```diff
@@ -29,6 +29,22 @@ let
+    netbootxyz = if cfg.netbootxyz.enable then pkgs.netbootxyz-efi else "";
+
+    copyExtraFiles = pkgs.writeShellScript "copy-extra-files" ''
+      empty_file=$(mktemp)
```
nitpick if you're up to it: we've since learned about the `pkgs.emptyFile` trivial builder. might be better to use that, and solve the TMPDIR problem with the installer?

apart from that, LGTM

Enzime replied:
https://github.com/NixOS/nixpkgs/blob/1783cfde53a58b1fa38449e47f45dd65091de7ae/nixos/modules/installer/tools/nixos-enter.sh#L103-L104

I think the `TMPDIR` needs to be fixed irrespective of whether this PR uses it or not, as the `nixos-enter` script creates `/tmp` but then won't even use it if `TMPDIR` is set.

My current preference is to keep `mktemp` to dogfood the `nixos-enter` command :)

pennae replied:
how did you find that? just tested an installer run (on 21.11 though, not unstable) and found that TMPDIR wasn't set.

pennae replied:
interactions between different sides of the same chroot sure can be confusing. :+1:
